### PR TITLE
Fix some entity mixins not storing their state

### PR DESCRIPTION
**Pull request by simon816**
Initially discovered in https://github.com/SpongePowered/Sponge/pull/165#discussion_r24569829.

Entity mixins that define additional data do not save this data to the disk since vanilla is not aware it.
I have gone through every entity mixin to find any data that will not persist.
It is unlikely for a lightning bolt to need to be persistent but I have added it anyway.

The following pull requests may have some data that should be made persistent.
~~#126~~, ~~#165~~, ~~#166~~, #167, ~~#169~~ 
I can either wait for the PRs to be pulled, then update this one to fix the issues. Or pull this and I can suggest changes needed for the other PRs.

### CODE REVIEW

**Comment by Mumfrey** [comment]
This looks good. Are the additional data supposed to be stored in the NBT though? The problem with storing additional data in the NBT is that if the world is loaded in a non-sponge implementation and the entity is read then written then the additional data are lost.

Would it instead be worthwhile to leverage the serialisation API instead and store all the additional data in a separate folder tree? Decoupling the custom data would preserve them across loads where Sponge isn't present.

In the days of CraftBukkit, the chances of this happening were pretty slim unless the server owner specifically launched vanilla (like when a new minecraft came out and a server owner wanted to skip ahead and use vanilla until the new CB came out), but since Sponge is a mod there are other circumstances that could cause it to not be loaded and the user to not know about it until it's too late and the damage is already done:

For example, the user is running other coremods and one of them interferes with the loading of Sponge, Sponge doesn't load and the user merrily loads up their world and kills all the extra data for entities in loaded chunks. Or a user running Sponge client-side upgrades their forge installation to a version not supported by the version of sponge they have installed and the same thing happens.

Worth thinking about. Since we don't control the vertical like CraftBukkit did, we can't make the same assumptions unfortunately.

**Comment by AIOOB** [comment]
This is our conversation with @bloodmc: [Chat log](http://sbnc.khobbits.co.uk/log/logs/old/spongedev_%5B2015-02-12%5D.htm)
Although we didn't really consider Sponge not being loaded, it has some interesting points in there.

**Comment by Mumfrey** [comment]
@AIOOB it looks from that conversation like @simon816 suggested the sensible idea of putting the data under `ForgeData` so that the data aren't lost at least if Forge is present. Since this is a Forge-based implementation that would make sense and would address the two issues I mentioned above (if Forge does indeed preserve keys in that tag even if a specific mod isn't present) but then this seems to counteract it

> <AIOOB> simon816: Is it really neccesary? It is probably a good thing to be less reliant on forge maintaining the same structure.
> <simon816> Doesn't matter really, just thought that method could be used
> <@blood|work> simon816: I prefer it to be seperate NBT
> <@blood|work> not within Forge
> <simon816> OK
> <@blood|work> SpongeData sounds good

which opens the original problem right back up again, but without a technical reason but seemingly for reasons of neatness, which doesn't seem like a good reason to abandon a more reliable solution.

Reading up on Forge's mechanisms, the standard suggestion seems to be calling `Entity::getEntityData()`` which returns the underlying `ForgeData` compound tag. This seems to address @bloodmc's concern of *"forge maintaining the same structure* since having a reliance on (what seems to be) a fairly core method for serialisation has a very different flavour to relying on an implementation detail, so if Forge does change the underlying storage in the future it won't be a breaking change.

Calling `getEntityData` and then storing everything in a `SpongeData` tag within it seems like it would fix the sponge-possibly-not-being-loaded issue, provided Forge is always in use (which is a more reasonable assumption).

**Comment by bloodmc** [comment]
@Mumfrey brings up a good point about loading in a non-sponge environment. As Sponge is technically just another mod, it should be placed with all other mod data.
@simon816 Please store all Sponge data in a compound labeled "SpongeData" within the ForgeData compound

**Comment by simon816** [comment]
Done.

I think that if Sponge didn't load successfully, it should force stop the server in order to avoid things going wrong and make sure plugins don't run. Or should block the server from loading until the user decides whether it should continue (Just like FML has the `/fml confirm`)

File: src/main/java/org/spongepowered/mod/entity/projectile/ProjectileSourceSerializer.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.UUID;
+
+public class ProjectileSourceSerializer {
+
+    public static NBTBase toNbt(ProjectileSource projectileSource) {
+        if (projectileSource instanceof Entity) {
+            return new NBTTagString(((Entity) projectileSource).getUniqueId().toString());
+        }
+        if (projectileSource instanceof BlockProjectileSource) {
+            return new NBTTagLong(VecHelper.toBlockPos(((BlockProjectileSource) projectileSource).getBlock().getPosition()).toLong());
```
Comment (gabizou) [inline src/main/java/org/spongepowered/mod/entity/projectile/ProjectileSourceSerializer.java]: Just out of curiosity, wouldn't this be better handled with `ProjectileSource#toContainer()` of which we can serialize the container internally? Regardless how it's actually serialized?

Perhaps a `TODO` for future persistence purposes.

File: src/main/java/org/spongepowered/mod/mixin/entity/projectile/MixinEntityFishHook.java [hunk 2]
```diff
@@ -206,4 +209,18 @@ public void setFishingRodItemStack(ItemStack fishingRod) {
         this.fishingRod = fishingRod;
     }
 
+    @Override
+    public void readFromNbt(NBTTagCompound compound) {
+        // TODO super.readFromNbt(compound);
+        if (compound.hasKey("damageAmount")) {
```
Comment (kenzierocks) [inline src/main/java/org/spongepowered/mod/mixin/entity/projectile/MixinEntityFishHook.java]: This is never written.

Comment (simon816) [inline src/main/java/org/spongepowered/mod/mixin/entity/projectile/MixinEntityFishHook.java]: Oops, must have accidentally removed the write when playing about with `ProjectileSourceSerializer`. I'll fix this

File: src/main/java/org/spongepowered/mod/entity/projectile/ProjectileSourceSerializer.java (hunk 1 above)
Comment (simon816) [inline src/main/java/org/spongepowered/mod/entity/projectile/ProjectileSourceSerializer.java]: Good thought. I'll add a TODO

**Comment by Mumfrey** [comment]
> I think that if Sponge didn't load successfully, it should force stop the server in order to avoid things going wrong and make sure plugins don't run. Or should block the server from loading until the user decides whether it should continue (Just like FML has the /fml confirm)

You're assuming that Sponge loads in some capacity, but I think much more likely scenarios for failure are that sponge simply isn't present or doesn't get loaded at all somehow. It's very likely that if Sponge loads *partially* then the game will just crash anyway, which isn't a problem. I don't think there's any need to over-complicate the startup process by trying to define what "loaded successfully" actually means and trying to somehow measure it.

File: src/main/java/org/spongepowered/mod/mixin/entity/projectile/MixinEntityArrow.java [hunk 3]
```diff
@@ -111,4 +114,16 @@ public void setCritical(boolean critical) {
     public MixinEntityArrow(World worldIn) {
         super(worldIn);
     }
+
+    @Override
+    public void readFromNbt(NBTTagCompound compound) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/mod/mixin/entity/projectile/MixinEntityArrow.java]: I'm fairly sure we can use at annotation to mix in to the target method

Comment (simon816) [inline src/main/java/org/spongepowered/mod/mixin/entity/projectile/MixinEntityArrow.java]: @gabizou It does, [see here](https://github.com/SpongePowered/Sponge/pull/173/files#diff-d1e54a67a6e6054ffea2c372f361a87cR394)

**Comment by ST-DDT** [comment]
Maybe we should state in the javadocs that custom data arent shared between different Sponge Server impls.

File: src/main/java/org/spongepowered/mod/interfaces/IMixinEntity.java [hunk 4]
```diff
@@ -35,4 +36,9 @@
     Entity getTeleportVehicle();
 
     void setTeleportVehicle(Entity entity);
+
+    void readFromNbt(NBTTagCompound compound);
```
Comment (bloodmc) [inline src/main/java/org/spongepowered/mod/interfaces/IMixinEntity.java]: This is not needed. Use the supermixins branch in Mixin then use SoftOverride annotation instead.

**Comment by bloodmc** [comment]
Please refactor this PR using the feature/supermixin branch in Mixin.

**Comment by simon816** [comment]
I've updated it to use the supermixin functionality.
There's a few javadoc comments in `MixinEntity` to explain what the different methods are for.

**Comment by simon816** [comment]
I've added the same methods to `MixinTileEntity` too, even though looking through the current tile entity mixins there's no issues, it's there for future use.

**Pull request closed without merging**